File: cerkefs/src/CerkeFS/Internal/Piece2.hs
module CerkeFS.Internal.Piece2
(Color(..)
,Side(..)
,Piece(..)
,Profession(..)
) where
data Color
 = Kok1  -- ^ Red, 赤
 | Huok2 -- ^ Black, 黒
 deriving(Show, Eq, Ord)
data Side
 = Upward -- ^ Pieces that points upward. Denoted by @^@ in the ASCII notation.
 | Downward-- ^ Pieces that points downward. Denoted by @_@ in the ASCII notation.
 deriving(Show, Eq, Ord)
data Piece 
 = Tam2 -- ^ Minds, 皇, tam
 | Piece {
  color :: Color, -- ^ The color of the piece
  prof :: Profession, -- ^ The profession of the piece
  side :: Side -- ^ The side that the piece belongs to
 } deriving(Show, Eq, Ord)
data Profession
 = Nuak1 -- ^ Vessel, 船, felkana  
 | Kauk2 -- ^ Pawn, 兵, elmer
 | Gua2  -- ^ Rook, 弓, gustuer
 | Kaun1 -- ^ Bishop, 車, vadyrd
 | Dau2  -- ^ Tiger, 虎, stistyst
 | Maun1 -- ^ Horse, 馬, dodor  
 | Kua2  -- ^ Clerk, 筆, kua 
 | Tuk2  -- ^ Shaman, 巫, terlsk  
 | Uai1  -- ^ General, 将, varxle
 | Io    -- ^ King, 王, ales
 deriving(Show, Eq, Ord, Enum)

File: cerke_fs/utils/src/piece2.rs
use piece2::Piece::NonTam2Piece;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Profession {
    /// Vessel, 船, felkana
    Nuak1,
    /// Pawn, 兵, elmer
    Kauk2,
    /// Rook, 弓, gustuer
    Gua2,
    /// Bishop, 車, vadyrd
    Kaun1,
    /// Tiger, 虎, stistyst
    Dau2,
    /// Horse, 馬, dodor  
    Maun1,
    /// Clerk, 筆, kua
    Kua2,
    /// Shaman, 巫, terlsk  
    Tuk2,
    /// General, 将, varxle
    Uai1,
    /// King, 王, ales
    Io,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Color {
    /// Red, 赤
    Kok1,
    /// Black, 黒
    Huok2,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Side {
    /// Pieces that points upward. Denoted by @^@ in the ASCII notation.
    Upward,
    /// Pieces that points downward. Denoted by @_@ in the ASCII notation.
    Downward,
}

#[derive(Debug, Clone)]
pub enum Piece {
    /// Minds, 皇, tam
    Tam2,
    NonTam2Piece {
        /// The color of the piece
        color: Color,
        /// The profession of the piece
        prof: Profession,
        /// The side that the piece belongs to
        side: Side,
    },
}

pub fn to_phantom(piece: &Piece) -> Option<PhantomPiece> {
    match piece {
        NonTam2Piece {
            color: c,
            prof: p,
            side: s,
        } => Some((c.clone(), p.clone(), s.clone())),
        _ => None,
    }
}

pub type PhantomPiece = (Color, Profession, Side);
